disable kthvalue overlap

guol-fnst opened the pull request:
Fixes #47934

guol-fnst commented:
Hi @kurtamohler 
Could you please help to review this PR?
I am not entirely  sure whether it is write.

ngimel reviewed (commented):
You should also check for overlap on the gpu

glaringlee commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -216,6 +217,10 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cpu(
   // FIXME: This seems bogus, I only do this because it was the old behaviour.
   //        The reductions are fine, as long as the axis being reduced along
   //        isn't of 0 elements (and the output has elements).
+  auto lap = at::get_overlap_status(self, values);
```
Can you put your code after the below '0 element check'? The above comments is related to '0 element check' IMO, we'd better keep them together.

guol-fnst replied:
Done!

kurtamohler commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -216,6 +217,10 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cpu(
   //        The reductions are fine, as long as the axis being reduced along
   //        isn't of 0 elements (and the output has elements).
+  auto lap = at::get_overlap_status(self, values);
+  TORCH_CHECK(lap != at::MemOverlapStatus::PARTIAL &&
```
Perhaps it would be good to move this logic to a new function `void assert_no_overlap(const Tensor& a, const Tensor& b)` in `MemoryOverlap.h/cpp`, since we'll need to perform this check in at least one other place (cuda)

guol-fnst replied:
Done!

kurtamohler commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -216,6 +217,10 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cpu(
+  auto lap = at::get_overlap_status(self, values);
+  TORCH_CHECK(lap != at::MemOverlapStatus::PARTIAL &&
+                lap != at::MemOverlapStatus::FULL,
+                "Found overlap in output");
```
I think this message should be a bit more clear. Maybe something like "output tensor's memory overlaps with input tensor"

guol-fnst replied:
Done！

guol-fnst commented:
> You should also check for overlap on the gpu

GPU check has been added.Thanks for the advice!

kurtamohler commented on test/test_torch.py:
```diff
@@ -9040,6 +9040,17 @@ def test_kthvalue(self, device, dtype):
 
+    # test overlapping output
+    @dtypes(torch.double)
+    def test_kthvalue_overlap(self, device, dtype):
```
I think this needs the `@onlyOnCPUAndCUDA` decorator, to avoid running on the XLA CI job, since this test is failing there.

guol-fnst replied:
Done！

ngimel commented on test/test_torch.py:
```diff
@@ -9040,6 +9040,18 @@ def test_kthvalue(self, device, dtype):
+        S = 10
+        k = 5
+        a = torch.randn(S)
+        values = torch.empty_like(a)
```
you don't need `values`

guol-fnst replied:
Done! 
Besides, due to the rebase conflicts, I merge the last  serveral commits to only one and forced a push to branch.
Thanks for the review!

Commit:
disable kthvalue overlap

VitalyFedyunin commented:
Bunch on onnx test failures seems to be unrelated, can you please rebase PR.

guol-fnst commented:
> Bunch on onnx test failures seems to be unrelated, can you please rebase PR.

I have rebased the branch, and everything seems ok!
Thanks!

guol-fnst commented:
Hi @VitalyFedyunin , Could you help to land this PR?
Thanks!

VitalyFedyunin commented:
Hi! Csn you please rebase it. I'm getting `redefinition of 'assert_no_overlap'` clangtidy error.

guol-fnst commented:
> Hi! Csn you please rebase it. I'm getting `redefinition of 'assert_no_overlap'` clangtidy error.

Issue #48651  has defined a similar function "assert_no_overlap" ,whichi merged on  Dec 10.
So I removed my defination and rebased it again.
Thanks!

guol-fnst committed:
remove redefined func
